feat: `aggsender` docs update (#255)
agglayer/aggkit: docs/aggsender.md

    Aggsender->>Agglayer: Send certificate
```

### PessimisticProof Mode

`Aggsender` will wait until the epoch event is triggered and ask the `L2BridgeSyncer` if there are new bridges and claims to be sent to `Agglayer`. Once we reach the moment in epoch when we need to send a certificate, the `Aggsender` will poll all the bridges and claims from the bridge syncer, based on the last sent L2 block to the `Agglayer`, until the block that the syncer has.

It is important to mention that no certificate will be sent to the `Agglayer` if the syncer has no bridges, since bridges change the Local Exit Root (`LER`).

If we have bridges, certificate will be built, signed, and sent to the `Agglayer` using the provided `Agglayer` RPC URL.

> Currently, `Agglayer` only supports one certificate per L1 epoch, per network, so we can not send more than one certificate. After the certificate is sent, we wait until the next epoch, either to resend it if its status is `InError`, or to build a new one if its status `Settled`. Also, we have no limit yet in how many bridges and claims can be sent in a single certificate. This might be something to test and check, because certificates carry a lot of data through RPC, so we might hit the rpc layer limit at some point. For this reason, we introduced the `MaxCertSize` configuration parameter on the `Aggsender`, where the user can define the maximum size of the certificate (based on the rpc communication layer limit) in bytes, and the `Aggsender` will limit the number of bridges and claims it will send to the `Agglayer` based on this parameter. Since both bridges and claims carry fixed size of data (each field is a fixed size field), we can we great precision calculate the size of a certificate.

Reviewer: Currently there are a  check of last certificate submitted for an early retry: https://github.com/agglayer/aggkit/blob/4ebab70b44d95e1fd7e253dfc671c1945d9d8c8e/aggsender/aggsender.go#L180

Author: It is mentioned in the paragraph that starts on line 43.

`InError` status on a certificate can mean a number of things. It can be an error that happened on the `Agglayer`. It can be an error in the data `Aggsender` sent, or the certificate was sent in between two epochs, which `Agglayer` considers invalid. Either way, the given certificate needs to be re-sent in the next epoch (or immediately after we notice its status change based on the `RetryCertAfterInError` config parameter), with all the previously sent bridges and claims, plus the new ones that happened after them, that the syncer saw and saved.

It is important to mention that, in the case of resending the certificate, the certificate height must be reused. If we are sending a new certificate, its height must be incremented based on the previously sent certificate.

Suppose the previously sent certificate was not marked as `InError`, or `Settled` on the `Agglayer`. In that case, we can not send/resend the certificate, even though a new epoch event is handled since it was not processed yet by the `Agglayer` (neither `Settled` nor marked as `InError`).

> The image below, depicts the interaction between different components when building and sending a certificate to the `Agglayer` in `PessimisticProof` mode.

Reviewer 2: ```suggestion
The image below, depicts the interaction between different components when building and sending a certificate to the `Agglayer` in the `PessimisticProof` mode.
```

    participant AggLayer
    participant L2BridgeSyncer
    participant L1InfoTreeSync
    participant AggSender

    User->>L1RPC: bridge (L1->L2)
    L1RPC->>Bridge: bridgeAsset
    Bridge->>AggLayer: updateL1InfoTree
    Bridge->>Bridge: auto claim

    User->>L2RPC: bridge (L2->L1)
    L2RPC->>L2BridgeSyncer: bridgeAsset emits bridgeEvent

>     User->>L2RPC: claimAsset

Reviewer 2: ```suggestion
    User->>L2RPC: claimAsset emits claimEvent
```

>     L2RPC->>L1InfoTreeSync: claimEvent

Reviewer 2: ```suggestion
    L2RPC->>L1InfoTreeSync: index claimEvent
```

    AggSender->>L2BridgeSyncer: get published bridges
    AggSender->>L2BridgeSyncer: get imported bridge exits
    Note right of AggSender: generate a Merkle proof for each imported bridge exit
    AggSender->>L1InfoTreeSync: get l1 info tree merkle proof for imported bridge exits
    AggSender->>AggLayer: send certificate
```

### AggchainProof mode

In essence, the `AggchainProof` mode follows the same logic and flow as `PessimisticProof` mode. Only difference is in two points:
- calling the `aggchain prover` to generate an `aggchain proof` that will be sent in the certfiicate to the `Agglayer`.
- resending an `InError` certficate does not expand it with new bridges and events that the syncer might have gotten in the meantime. This is done because `aggchain prover` already generated a proof for a given block range, and since proof generation can be a long process, this is a small optimization. Note that 
this might change in the future.

Calling the `aggchain prover` is done right before signing and sending the certificate to the `Agglayer`. To generate an `aggchain proof` prover needs couple of things:
- block range on L2 for which we are trying to generate a certificate.

> - finalized L1 info tree root and its leaf and proof on the L1 info tree. Basically, this is the latest finalized l1 info tree root needed by the prover to generate the proof. This root is also use to generate merkle proof for every imported bridge exit (claim) in certificate.

Reviewer 2: ```suggestion
- finalized L1 info tree root, leaf, and proof on the L1 info tree. Basically, this is the latest finalized l1 info tree root needed by the prover to generate the proof. This root is also use to generate merkle proof for every imported bridge exit (claim) in certificate.
```

- injected GlobalExitRoot's on L2 and their leaves and proofs. Merkle proofs of the injected GERs are calculated based on the finalized L1 info tree root.
- imported bridge exits (claims) we intend to include in the certificate for the given block range.

> The image below, depicts the interaction between different components when building and sending a certificate to the `Agglayer` in `PessimisticProof` mode.

Reviewer 2: ```suggestion
The image below, depicts the interaction between different components when building and sending a certificate to the `Agglayer` in the `AggchainProof` mode.
```

| Field Name               | Description                                                                 |
|--------------------------|-----------------------------------------------------------------------------|
| `network_id`               | This is the id of the rollup (>0)                                                       |
| `height`                   | Order of certificates. First one is 0                                                   |

> | `prev_local_exit_root`     | The first one must be the one in SMC (currently is a 0x000…00)                          |

Reviewer 2: ```suggestion
| `prev_local_exit_root`     | The first one must be the one in smart contract (currently is a 0x000…00)                          |
```

| `new_local_exit_root`      | It’s the root after bridge_exits                                                        |
| `bridge_exits`             | These are the leaves of the LER tree included in this certificate. (bridgeAssert calls) |
| `imported_bridge_exits`    | These are the claims done in this network                                               |
| `aggchain_params`          | Aggchain params returned by the aggchain prover                                         |
| `aggchain_proof`           | Aggchain proof generated by the aggchain prover                                         |
| `custom_chain_data`        | Custom chain data returned by the aggchain prover                                       |

## Configuration

| Name                          | Type               | Description                                                                                                     |
|-------------------------------|--------------------|-----------------------------------------------------------------------------------------------------------------|

> | StoragePath                   | string             | Path where to store Aggsender DB                                                                     |

Reviewer: This field is the path and the filename for database file

Author: Fix: https://github.com/agglayer/aggkit/pull/255/commits/0fcbfca7254172b1f00b10a3464e564d573aaf53

## Additional Documentation

> [1] AggLayer integration guide_WIP

Reviewer 2: Is there any URL that hosts this guide, so that we can provide it here?

Author: No, I will remove this line for now.
